frame/support/src/executor.rs

```diff
@@ -0,0 +1,620 @@
+/// this trait is concerned, it can be now, all at once. It can be executed as mandatory work in
+/// `on_initialize` or `on_finalize`, or in some other low priority circumstance (e.g. on_idle).
+///
+/// If the type implementing this trait is generic over `<T: Config>` then one needs to derive
+/// [`CloneNoBound`], [`PartialEqNoBound`], [`EqNoBound`], [`RuntimeDebugNoBound`], as opposed to
+/// their normal counterparts, and implement `Default` manually. This is to prevent `T` to be
+/// bounded to these traits.
+pub trait RuntimeTask:
+	Sized + Clone + Default + Encode + Decode + PartialEq + Eq + sp_std::fmt::Debug
+{
+	/// Execute the task while consuming self. The task must not most consume more than `max_weight`
+	/// under any circumstance. Consuming *less* than `max_weight` is allowed.
+	///
+	/// A tuple is returned, where the items are as follows:
+	///   1. Option<Self>, where `None` means that this task is now complete (and shall not be kept
+	///      in storage anymore), and `Some(_)` indicating that this task is not yet complete, and
+	///      should be executed at a later time.
+	///   2. The actual amount of weight that was consumed. Must always be less than `max_weight`.
+	///      parameter.
+	///
+	/// It is critically important for a task to only return a non-zero consumed weight **ONLY if it
+	/// _actually did something_**. If a positive weight is returned, then an executor could
+	/// interpret this as a task that could use another execution slot, and continue the execution
+	/// potentially for numerous iterations.
+	fn execute(self, max_weight: Weight) -> (Option<Self>, Weight);
+
+	/// The leftover weight that this task expects to execute, if any.
+	#[cfg(test)]
+	fn leftover(&self) -> Weight;
+}
+
+#[cfg(any(test, feature = "std"))]
+impl RuntimeTask for () {
+	fn execute(self, _: Weight) -> (Option<Self>, Weight) {
+		(None, 0)
+	}
+	#[cfg(test)]
+	fn leftover(&self) -> Weight {
+		0
+	}
+}
+
+/// Common trait for a executor that is stored as a storage item.
+pub trait StoredExecutor {
+	/// The task type used by this executor.
+	type Task: RuntimeTask;
+
+	/// Something that can define how much weight quote this executor is allowed to use per
+	/// execution.
+	type Quota: Get<Weight>;
+
+	/// Execute all tasks based on an unspecified strategy, consuming at most `Self::Quota` and
+	/// returning the actual amount of weight consumed.
+	///
+	/// The returned weight must take into account the cost of internal operations of the
+	/// implementation, such as scheduling, as well. But, it DOES NOT take into account any
+	/// potential storage operations that needed to be performed to fetch `Self` from storage.
+	///
+	/// A  sensible patter of using an implementation of this trait is therefore:
+	///
+	/// ```ignore
+	/// let mut consumed = <StorageItemExecutor<T>>::mutate(|e| e.execute());
+	/// consumed += <weight of 1 read and write for the mutate call above>;
```

Comment:
This disregards encoding/decoding weight.

Reply:
Since I am aware of this and there's a TODO for it, I'll resolve this to keep the page tidy.